### PR TITLE
feat: First draft of parsing .eml files in the external layer

### PR DESCRIPTION
**Pull request by jens1o**

### CODE REVIEW

**Comment by jens1o** [comment]
@Jonas-Karl could you please take a look? I'm keen to hear your feedback. If I see this correctly, I have implemented the external *and* the gateway part of this epic.

I have also introduced a new dependency that does the heavy lifting (parsing according to the RFC 822). Is this fine for you?

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/domain/email/Email.java [hunk 1]
```diff
@@ -17,7 +17,7 @@ public final class Email extends AggregateRoot<EmailId> {
     private String content;
     private Set<Attachment> attachments;
 
-    private Email(EmailId id, String content, EmailMetadata emailMetadata,
+    public Email(EmailId id, String content, EmailMetadata emailMetadata,
```
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/domain/email/Email.java:20]: Why did you change this? There are factory `create` factory methods

**Review by Jonas-Karl — COMMENTED** [review]
Just a quick look. I'll take a bit more time in the next days

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/domain/email/Email.java (hunk 1 above)
Comment (jens1o) [reply src/main/java/de/dhbw/karlsruhe/students/mailflow/core/domain/email/Email.java:20]: oh I didn't noticed 🤣 

great catch!

**Review by Jonas-Karl — CHANGES_REQUESTED** [review]
Do you want to correctly implement everything in this PR (and make it bigger) or do you want to finish everything by fixing the smaller issues and then rearrange the code afterwards in a smaller ticket? I prefer the 2nd alternative. What about you?

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/EmailMetadataFactory.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private List<Address> getRecipients(RecipientType recipientType) throws MessagingException {
+        List<Address> addresses = new ArrayList<>();
+
+        jakarta.mail.Address[] recipients = message.getRecipients(recipientType);
+
+        if (recipients == null || recipients.length == 0) {
+            return Collections.emptyList();
+        }
```
[the remark this answers is not in the document]
Comment (jens1o) [reply src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/EmailMetadataFactory.java:36]: the `recipients == null` check is necesary to avoid a NPE, the `recipients.length == 0` is just a early return

File: src/test/java/de/dhbw/karlsruhe/students/mailflow/ParseEmlFIleTest.java
[the remark this answers is not in the document]
Comment (jens1o) [reply src/test/java/de/dhbw/karlsruhe/students/mailflow/ParseEmlFIleTest.java:1]: aww thank you :33333

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/domain/email/Email.java
[the remark this answers is not in the document]
Comment (jens1o) [reply src/main/java/de/dhbw/karlsruhe/students/mailflow/core/domain/email/Email.java:1]: What do you mean? Do you have an example where I could use the external layer here?

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/Parser.java
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/Parser.java:1]: Move this file to the usecase layer (create a folder at src/usecase), rename it to EmlParser and remove the dependency to jakarta (the exception). If needed create an own exception. Im not sure where to put it. Probably in the domain layer.

Comment (jens1o) [reply src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/Parser.java:1]: it's an interface?

Comment (jens1o) [reply src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/Parser.java:1]: But sure, I can make an abstraction over jakarta :)

Comment (Jonas-Karl) [reply src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/Parser.java:1]: Not sure what you mean with "its an interface".

I guess we need to talk a bit more about what I mean on monday :)

File: src/test/java/de/dhbw/karlsruhe/students/mailflow/ParseEmlTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        InputStream inputStream = new ByteArrayInputStream("""
+                From: someone@example.com
+                To: someone_else@example.com
+                Subject: An RFC 822 formatted message
+
+                This is the plain text body of the message. Note the blank line
+                between the header information and the body of the message.""".getBytes(StandardCharsets.UTF_8));
+
+        // Act
+        EmlParser parser = new EmlParser();
```
Comment (Jonas-Karl) [inline src/test/java/de/dhbw/karlsruhe/students/mailflow/ParseEmlTest.java:35]:
```suggestion
        EmailParser parser = new EmlParser();
```
use interface

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/EmlParser.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import jakarta.mail.internet.MimeMessage;
+
+/**
+ * Implements "RFC822: Standard for ARPA Internet Text Messages"
+ * by parsing .eml-files
+ */
+public class EmlParser implements EmailParser {
+    private InputStream inputStream;
+
+    public Email parseToEmail() throws EmailParsingException {
```
Comment (Jonas-Karl) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/EmlParser.java:21]:
```suggestion
    @Override
    public Email parseToEmail() throws EmailParsingException {
```

File: src/test/assets/eml_files/microsoft-example-eml-message.eml
Comment (Jonas-Karl) [inline src/test/assets/eml_files/microsoft-example-eml-message.eml:1]: can this be deleted?

**Review by Jonas-Karl — CHANGES_REQUESTED** [review]

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/EmailMetadataFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jonas-Karl) [reply src/main/java/de/dhbw/karlsruhe/students/mailflow/core/external/email/EmailMetadataFactory.java:35]: What I mean is, you could implement this functionality in the Email aggregate and this functionality inside the domain. I think its part of the domain logic to return an emptyList when no recipients exist.

Anyways, this implementation would need a bit more to be adjusted, because here we are working with jakarta emails but we could do this with our domain emails. We can change it afterwards if you want. Also I am not 100% sure if it works the way I imagine it

**Review by Jonas-Karl — CHANGES_REQUESTED** [review]

**Review by Jonas-Karl — CHANGES_REQUESTED** [review]

File: src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/email/EmailMetadataFactory.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class EmailMetadataFactory {
+    private Message message;
+
+    private EmailMetadataFactory(Message message) {
+        if(message == null){
```
Comment (jens1o) [inline src/main/java/de/dhbw/karlsruhe/students/mailflow/external/infrastructure/email/EmailMetadataFactory.java:16]:
```suggestion
        if (message == null){
```

File: src/test/assets/eml_files/microsoft-example-eml-message.eml
Comment (jens1o) [inline src/test/assets/eml_files/microsoft-example-eml-message.eml:1]: could you please delete the file? :)

**Review by jens1o — COMMENTED** [review]
lgtm, just a few final tweaks necessary :)

**Comment by Jonas-Karl** [comment]
What about missing bad-case tests? #74

**Comment by jens1o** [comment]
I have written a first test-case for a negative sample. Is this fine for you? :)

**Review by Jonas-Karl — APPROVED** [review]
